[Refactor]Rename NCCL-related items to comm_backend

noemotiovon opened the pull request:
## Why are these changes needed?

### Background
This PR is a follow-up to [#51032](https://github.com/ray-project/ray/pull/51032), which introduced multi-device support in the Compiled Graph by leveraging CUDA's NCCL backend for efficient out-of-band tensor communication.

While the current implementation is tightly coupled with NCCL and CUDA, the Compiled Graph runtime is now ready to support a broader spectrum of device types and collective communication backends (e.g., HCCL, RCCL).

### What This PR Does?
To enable extensibility and backend-agnostic design, this PR introduces the following core changes:

Refactored NCCL-specific naming and APIs
NCCL-related modules, classes, and function names have been generalized to eliminate hardcoded CUDA/NCCL assumptions.

Introduced a pluggable communication backend interface
A unified abstraction layer is added to decouple collective communication logic from any specific implementation. This makes it easier to support alternative collective libraries and device types in the future.

This refactor does not alter the existing behavior of NCCL-based Compiled Graph execution. All current workflows using CUDA+NCCL continue to function as before.


## Related issue number
https://github.com/ray-project/ray/issues/51574

<!-- For example: "Closes #1234" -->

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a
           method in Tune, I've added it in `doc/source/tune/api/` under the
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

noemotiovon commented:
@ruisearch42, 
Hi, It’s possible that in the future, the communication backend will not be limited to NCCL, as shown in this [PR](https://github.com/ray-project/ray/pull/51032) with HCCL. Therefore, would it be possible to adjust the naming of NCCL?

noemotiovon commented:
This PR need follow https://github.com/ray-project/ray/issues/51574. Make it draft.

noemotiovon committed:
[Compiled Graph] Refactor NCCL Integration for Extensibility

This commit is a follow-up to #51032, which introduced multi-device support
in the Compiled Graph by leveraging CUDA's NCCL backend for efficient out-of-band
tensor communication. While the current implementation is NCCL-specific, the
Compiled Graph runtime is now ready to support a broader range of device types
and collective communication libraries.

To prepare for this generalization, this commit introduces the following changes:

1. Refactored NCCL-specific naming and interfaces
2. Established a pluggable communication backend interface

This refactor does not change the behavior for existing NCCL-based Compiled Graph
execution, but lays the foundation for enabling collective communication across
diverse hardware accelerators and runtime environments.

Signed-off-by: noemotiovon <757486878@qq.com>

noemotiovon committed:
fix ci issue

Signed-off-by: noemotiovon <757486878@qq.com>

noemotiovon committed:
fix test issue

Signed-off-by: noemotiovon <757486878@qq.com>

ruisearch42 commented on python/ray/dag/dag_node.py:
```diff
@@ -184,7 +185,14 @@ def with_tensor_transport(
             )
         elif transport == "nccl":
             self._type_hint = TorchTensorType(
-                transport=transport,
+                transport="accelerator",
+                device=device,
+                _static_shape=_static_shape,
+                _direct_return=_direct_return,
+            )
+        elif transport == "accelerator":
+            self._type_hint = TorchTensorType(
+                transport == "accelerator",
                 device=device,
                 _static_shape=_static_shape,
                 _direct_return=_direct_return,
```
Does "accelerator" mean non-GPU accelerator or it also includes GPU?

noemotiovon replied:
Yes, by "accelerator", we refer broadly to all hardware accelerators, including GPUs, NPUs, and others—so GPUs are certainly included.

We’ve currently kept the mention of NCCL in the comment, as with_tensor_transport is explicitly used by frameworks like vLLM, where GPU-based communication is common. That said, the underlying TorchTensorType already generalizes to any accelerator, and the transport layer is not limited to NCCL.

We’re happy to revisit or simplify this comment if others feel that referring to NCCL is too GPU-specific. Open to suggestions!

ruisearch42 replied:
SG. Can you clarify in the arg docs the relationship between "nccl" and "accelerator", and when to use which?
You can additionally mention "nccl" is for backwards compatibility.

ruisearch42 reviewed (commented):
Overall LGTM, initial reviews

noemotiovon commented on python/ray/experimental/channel/torch_tensor_type.py:
```diff
@@ -77,13 +78,16 @@ def __init__(
 
-        if transport not in [self.AUTO, self.NCCL, self.CPU]:
+        if transport not in [self.AUTO, self.CPU, self.ACCELERATOR]:
+            print(f"lcg ====>{transport=}")
```
Oops, looks like our secret debugging message made it into the commit :no_mouth:
Thanks for the catch

noemotiovon committed:
fix review comment

Signed-off-by: noemotiovon <757486878@qq.com>

noemotiovon commented:
Hi @ruisearch42，
Thanks again for your earlier review! I’ve updated the PR according to the feedback. When you have time, could you please take another look? Appreciate it!

ruisearch42 commented on python/ray/dag/dag_node_operation.py:
```diff
@@ -141,14 +141,15 @@ def __lt__(self, other: "_DAGOperationGraphNode"):
-            # the NCCL operation.
-            return self.is_nccl_op
+        if self.is_accelerator_op != other.is_accelerator_op:
+            # When one node is a accelerator operation and the other is not,
```
```suggestion
            # When one node is an accelerator operation and the other is not,
```

ruisearch42 commented on python/ray/dag/dag_node_operation.py:
```diff
@@ -178,9 +179,10 @@ def in_degree(self) -> int:
-        If a node is not a NCCL operation, it is ready when it has a zero in-degree.
-        If it is a NCCL operation, it is ready when all the nodes in the operation
-        have zero in-degrees.
+        If a node is not a accelerator operation, it is ready when it has a zero
```
```suggestion
        If a node is not an accelerator operation, it is ready when it has a zero
```

ruisearch42 commented on python/ray/dag/dag_node_operation.py:
```diff
@@ -178,9 +179,10 @@ def in_degree(self) -> int:
-        have zero in-degrees.
+        If a node is not a accelerator operation, it is ready when it has a zero
+        in-degree.
+        If it is a accelerator operation, it is ready when all the nodes in the
```
```suggestion
        If it is an accelerator operation, it is ready when all the nodes in the
```

ruisearch42 commented on python/ray/dag/dag_node_operation.py:
```diff
@@ -191,31 +193,42 @@ def is_read(self) -> bool:
+    def is_accelerator_read(self) -> bool:
         """
-        A node is a NCCL read if it is a read node and requires NCCL.
+        A node is a accelerator read if it is a read node and requires accelerator.
```
```suggestion
        A node is an accelerator read if it is a read node and requires accelerator.
```

ruisearch42 commented on python/ray/dag/dag_node_operation.py:
```diff
@@ -191,31 +193,42 @@ def is_read(self) -> bool:
+    def is_accelerator_compute(self) -> bool:
         """
-        A node is a NCCL compute if it is a compute node and requires NCCL.
+        A node is a accelerator compute if it is a compute node and requires accelerator.
```
```suggestion
        A node is an accelerator compute if it is a compute node and requires accelerator.
```

ruisearch42 commented on python/ray/dag/dag_node_operation.py:
```diff
@@ -191,31 +193,42 @@ def is_read(self) -> bool:
+    def is_accelerator_write(self) -> bool:
         """
-        A node is a NCCL write if it is a write node and requires NCCL.
+        A node is a accelerator write if it is a write node and requires accelerator.
```
```suggestion
        A node is an accelerator write if it is a write node and requires accelerator.
```

ruisearch42 commented on python/ray/dag/dag_node_operation.py:
```diff
@@ -320,18 +335,18 @@ def _select_next_nodes(
     When a node has a zero in-degree, it is added to the corresponding actor's
-    priority queue. For a node other than a NCCL collective node, it is ready to be
-    executed if it has a zero in-degree. For a NCCL collective node, it is ready to
+    priority queue. For a node other than a accelerator collective node, it is ready to be
```
```suggestion
    priority queue. For a node other than an accelerator collective node, it is ready to be
```

ruisearch42 commented on python/ray/dag/dag_node_operation.py:
```diff
@@ -320,18 +335,18 @@ def _select_next_nodes(
-    priority queue. For a node other than a NCCL collective node, it is ready to be
-    executed if it has a zero in-degree. For a NCCL collective node, it is ready to
+    priority queue. For a node other than a accelerator collective node, it is ready to be
+    executed if it has a zero in-degree. For a accelerator collective node, it is ready to
```
```suggestion
    executed if it has a zero in-degree. For an accelerator collective node, it is ready to
```

ruisearch42 commented on python/ray/dag/dag_node_operation.py:
```diff
@@ -320,18 +335,18 @@ def _select_next_nodes(
     be executed when all the nodes in its collective operation have zero in-degrees.
 
-    If a node is a NCCL collective node, it updates the `ready_collective_nodes` of
+    If a node is a accelerator collective node, it updates the `ready_collective_nodes` of
```
```suggestion
    If a node is an accelerator collective node, it updates the `ready_collective_nodes` of
```

ruisearch42 commented on python/ray/dag/dag_node_operation.py:
```diff
@@ -320,18 +335,18 @@ def _select_next_nodes(
 
-    If the selected node is a NCCL write node, select all the downstream NCCL
-    read nodes. If the selected node is a NCCL collective node, select all the NCCL
+    If the selected node is a accelerator write node, select all the downstream accelerator
```
```suggestion
    If the selected node is an accelerator write node, select all the downstream accelerator
```

ruisearch42 commented on python/ray/dag/dag_node_operation.py:
```diff
@@ -320,18 +335,18 @@ def _select_next_nodes(
-    If the selected node is a NCCL write node, select all the downstream NCCL
-    read nodes. If the selected node is a NCCL collective node, select all the NCCL
+    If the selected node is a accelerator write node, select all the downstream accelerator
+    read nodes. If the selected node is a accelerator collective node, select all the accelerator
```
```suggestion
    read nodes. If the selected node is an accelerator collective node, select all the accelerator
```

ruisearch42 commented on python/ray/dag/dag_node_operation.py:
```diff
@@ -729,10 +744,10 @@ def _generate_actor_to_execution_schedule(
         # Select a list of nodes to be executed. There are three cases:
-        # 1. If a selected node is not a NCCL operation, only itself is returned.
-        # 2. If a selected node is a NCCL write operation, the corresponding NCCL
+        # 1. If a selected node is not a accelerator operation, only itself is returned.
```
```suggestion
        # 1. If a selected node is not an accelerator operation, only itself is returned.
```

ruisearch42 commented on python/ray/dag/dag_node_operation.py:
```diff
@@ -729,10 +744,10 @@ def _generate_actor_to_execution_schedule(
-        # 1. If a selected node is not a NCCL operation, only itself is returned.
-        # 2. If a selected node is a NCCL write operation, the corresponding NCCL
+        # 1. If a selected node is not a accelerator operation, only itself is returned.
+        # 2. If a selected node is a accelerator write operation, the corresponding accelerator
```
```suggestion
        # 2. If a selected node is an accelerator write operation, the corresponding accelerator
```

ruisearch42 commented on python/ray/dag/dag_node_operation.py:
```diff
@@ -729,10 +744,10 @@ def _generate_actor_to_execution_schedule(
+        # 2. If a selected node is a accelerator write operation, the corresponding accelerator
         #    read operations are also returned.
-        # 3. If a selected node is a NCCL collective operation, all the nodes in
+        # 3. If a selected node is a accelerator collective operation, all the nodes in
```
```suggestion
        # 3. If a selected node is an accelerator collective operation, all the nodes in
```

ruisearch42 commented on python/ray/dag/dag_node_operation.py:
```diff
@@ -792,30 +807,30 @@ def _generate_overlapped_execution_schedule(
-                        # Found a NCCL read/write operation, skip the overlap
-                        # optimization to keep relative order of NCCL operations
+                    ) and overlapped_schedule[j].requires_accelerator:
+                        # Found a accelerator read/write operation, skip the overlap
```
```suggestion
                        # Found an accelerator read/write operation, skip the overlap
```

ruisearch42 commented on python/ray/experimental/channel/auto_transport_type.py:
```diff
@@ -11,8 +11,8 @@ class AutoTransportType(ChannelOutputType):
     With this type hint Compiled Graphs automatically decide the best transport
-    to use (e.g., NCCL or shared memory) based on the node locations and GPU IDs
-    of the readers and writers.
+    to use (e.g., Accellerator or shared memory) based on the node locations and
```
```suggestion
    to use (e.g., accelerator or shared memory) based on the node locations and
```

ruisearch42 commented on python/ray/experimental/channel/cpu_communicator.py:
```diff
@@ -91,7 +91,7 @@ def _apply_op(self, op: ReduceOp, tensors: List["torch.Tensor"]) -> "torch.Tenso
 class CPUCommunicator(Communicator):
     """
-    Uses a CPU-based communicator actor instead of a NCCL group.
+    Uses a CPU-based communicator actor instead of a communicator group like NCCL.
```
```suggestion
    Uses a CPU-based communicator actor instead of an accelerator group like NCCL.
```

ruisearch42 commented on python/ray/experimental/channel/torch_tensor_accelerator_channel.py:
```diff
@@ -401,39 +403,39 @@ def __init__(
+        # process is the driver, and there’s no need to fetch the comm_group.
         if self._typ.communicator_id in ctx.communicators:
-            self._nccl_group: "Communicator" = ctx.communicators[
+            self._comm_group: "Communicator" = ctx.communicators[
```
let's call this `_accelerator_group`? Please update the comments as well

ruisearch42 reviewed (approved):
Thanks for the PR. The raised issues are almost all nitpicks.

noemotiovon committed:
Fix review comment

Signed-off-by: noemotiovon <757486878@qq.com>

noemotiovon commented:
Hi @ruisearch42，
Thanks for the review, and apologies for the ambiguity caused by my oversight. I've addressed all the comments and corrected the issues accordingly. Looking forward to your continued review!

noemotiovon committed:
fix ci issue

Signed-off-by: noemotiovon <757486878@qq.com>

ruisearch42 commented:
multi-gpu test failed. Please take a look. @noemotiovon

noemotiovon commented:
Hi @ruisearch42 , 
Very sorry! I missed updating the assertions in the test file, which caused the CI failure. I've just fixed it—could you please help me re-trigger the CI? Thanks a lot!
Also, I saw you were working quite late—please don’t overwork yourself and make sure to get some rest. Really appreciate your help!  :heart:

noemotiovon committed:
Fix CI issue

Signed-off-by: noemotiovon <757486878@qq.com>

ruisearch42 commented:
@jjyao @edoakes could you help merge? thanks
